### DISCUSSION TITLE
does the React\Promise\all returns the results in order?

### DISCUSSION THREAD

**Discussion by Amir-adibi** [reactphp/reactphp#538] [Q&A] [answered]
Hello

I have a question about the `React\Promise\all` method.
I store the result of async calls of a method in an array and perform the `all()` method on it.
does the `all` method resolve these results in the order in which they were added to the array?

**Comment by clue** [comment, marked as answer]
> does the React\Promise\all returns the results in order?

Yes, the [`all()` function](https://github.com/reactphp/promise#all) is guaranteed to preserve the original array (or iterator) keys and their order.

```php
<?php

require __DIR__ . '/vendor/autoload.php';

React\Promise\all([
    'first' => new React\Promise\Promise(function (callable $resolve): void {
        React\EventLoop\Loop::addTimer(0.2, fn() => $resolve(1));
    }),
    'second' => new React\Promise\Promise(function (callable $resolve): void {
        React\EventLoop\Loop::addTimer(0.1, fn() => $resolve(2));
    })
])->then(function (array $results): void {
    var_dump($results);
});
```

```
array(2) {
  'first' =>
  int(1)
  'second' =>
  int(2)
}
```

https://play.phpsandbox.io/react/async/R6NgB5WAVgpx98z0

I hope this helps! :+1: If so, consider supporting this project, for example by [becoming a sponsor](https://github.com/sponsors/reactphp) ❤️
